zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+    String backupFilePath = constructBackupFilePath(srcName.getName());
+    File backupFile = new File(backupFilePath);
+    try {
+      inputStream = new FileInputStream(backupFile);
+    } catch (FileNotFoundException e) {
+      if (!backupFile.exists() || backupFile.isDirectory()) {
+        throw new BackupException("The backed up file with the file path " + backupFilePath
+            + " does exists or is a directory, could not copy to local storage.");
+      } else {
+        throw new IOException("Cannot open source file in backup storage" + backupFilePath
+            + " , cannot be copied to local storage.");
+      }
+    }
+
+    // Prepare in local storage: create parent dirs if not exist, create the destination file
+    destFile.getParentFile().mkdirs();
+    if (!destFile.createNewFile()) {
+      if (destFile.exists()) {
+        throw new BackupException("A local storage file with the file path " + destFile.getPath()
+            + " already exists, could not create a new one.");
+      } else {
+        throw new IOException("Failed to create a file at destination path: " + destFile.getPath()
+            + ", stopping the streaming.");
+      }
+    }
+
+    // Create output stream on the destination file and start streaming data
+    try {
+      outputStream = new FileOutputStream(destFile);
+
+      byte[] buffer = new byte[1024];
+      int lengthRead;
+      while ((lengthRead = inputStream.read(buffer)) > 0) {
+        outputStream.write(buffer, 0, lengthRead);
+        outputStream.flush();
+      }
+    } catch (FileNotFoundException e) {
+      throw new IOException(
+          "The destination local file " + destFile.getPath() + " could not be opened for write.");
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during data streaming while copying file " + backupFilePath
+              + " to local storage. The message of the IOException is: " + e.getMessage());
+    } finally {
+      inputStream.close();
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+  }
+
+  @Override
+  public void delete(File fileToDelete) throws IOException {
+    String backupFilePath = constructBackupFilePath(fileToDelete.getPath());
+    try {
+      if (!Files.deleteIfExists(Paths.get(backupFilePath))) {
+        throw new BackupException(
+            "File " + fileToDelete.getName() + "does not exist, cannot delete this file.");
+      }
+    } catch (DirectoryNotEmptyException e) {
+      throw new BackupException(
+          "Directory " + fileToDelete.getName() + "is not empty, cannot be deleted.");
+    }
+  }
+
+  /**
+   * Construct the file path in backup storage using file name
+   * @param fileName The file name for the file
+   * @return The file path to the location where the file will be stored in backup storage
+   */
+  private String constructBackupFilePath(String fileName) {
```

Comment:
This is something I mentioned above. Better have a public util for that. This is duplicate with BackupFileInfo constructor logic.

Reply:
Thanks, I think it's a good idea and I've moved this logic to a public util method. But it is not duplicated with BackupFileInfo constructor though, they do both try to determine the file type, but they return different things (file name prefix in this case, and file type in BackupFileInfo), so for now I'll keep both of them.